File: src/tables/vorg.cpp
#include "tables.h"

void parseVorg(Parser &parser)
{
    const auto majorVersion = parser.read<UInt16>("Major version");
    const auto minorVersion = parser.read<UInt16>("Minor version");
    if (!(majorVersion == 1 && minorVersion == 0)) {
        throw QString("invalid table version");
    }

    parser.read<Int16>("Default vertical origin");
    const auto count = parser.read<UInt16>("Number of metrics");
    parser.readArray("Metrics", count, [&](const auto index){
        parser.beginGroup(index);
        parser.read<GlyphId>("Glyph index");
        parser.read<Int16>("Coordinate");
        parser.endGroup();
    });
}

File: src/tables/vorg.rs
use crate::parser::*;
use crate::{Error, Result};

pub fn parse(parser: &mut Parser) -> Result<()> {
    let major_version: u16 = parser.read("Major version")?;
    let minor_version: u16 = parser.read("Minor version")?;
    if !(major_version == 1 && minor_version == 0) {
        return Err(Error::InvalidTableVersion);
    }

    parser.read::<i16>("Default vertical origin")?;
    let count = parser.read::<u16>("Number of metrics")?;
    for i in 0..count {
        parser.begin_group(format!("Metric {}", i));
        parser.read::<GlyphId>("Glyph index")?;
        parser.read::<i16>("Coordinate")?;
        parser.end_group();
    }

    Ok(())
}
